pkg/store/bucket.go

```diff
@@ -3159,10 +3161,11 @@ func (r *bucketIndexReader) loadSeries(ctx context.Context, ids []storage.Series
 			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+uint64(n+int(l)+1), bytesLimiter, tenant)
 		}
 		c = c[n : n+int(l)]
-		r.mtx.Lock()
+		r.loadedSeriesMtx.Lock()
 		r.loadedSeries[id] = c
+		r.loadedSeriesMtx.Unlock()
+
 		r.block.indexCache.StoreSeries(r.block.meta.ULID, id, c, tenant)
-		r.mtx.Unlock()
```

Comment:
This is a major change.
Previously the code was like below. `StoreSeries` can be time consuming and doesn't need to be protected by the lock.

```
r.mtx.Lock()
r.loadedSeries[id] = c
r.block.indexCache.StoreSeries(r.block.meta.ULID, id, c, tenant)
r.mtx.Unlock()
```